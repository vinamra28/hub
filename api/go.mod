module github.com/tektoncd/hub/api

go 1.14

require (
	github.com/Azure/azure-sdk-for-go v50.2.0+incompatible // indirect
	github.com/Azure/go-autorest/autorest v0.11.17 // indirect
	github.com/Azure/go-autorest/autorest/adal v0.9.10 // indirect
	github.com/Azure/go-autorest/autorest/to v0.4.0 // indirect
	github.com/Azure/go-autorest/autorest/validation v0.3.1 // indirect
	github.com/aws/aws-sdk-go v1.37.1 // indirect
	github.com/dgrijalva/jwt-go v3.2.0+incompatible
	github.com/docker/cli v20.10.2+incompatible // indirect
	github.com/docker/docker v20.10.2+incompatible // indirect
	github.com/emicklei/go-restful v2.15.0+incompatible // indirect
	github.com/go-gormigrate/gormigrate/v2 v2.0.0
	github.com/go-openapi/spec v0.20.2 // indirect
	github.com/go-testfixtures/testfixtures/v3 v3.2.0
	github.com/gogo/protobuf v1.3.2 // indirect
	github.com/google/go-containerregistry/pkg/authn/k8schain v0.0.0-20210129212729-5c4818de4025 // indirect
	github.com/google/go-github v17.0.0+incompatible
	github.com/google/gofuzz v1.2.0 // indirect
	github.com/googleapis/gnostic v0.5.3 // indirect
	github.com/hashicorp/go-version v1.2.0
	github.com/ikawaha/goahttpcheck v1.3.1
	github.com/joho/godotenv v1.3.0
	github.com/mitchellh/go-homedir v1.1.0
	github.com/sirupsen/logrus v1.7.0 // indirect
	github.com/spf13/cobra v1.1.1
	github.com/spf13/viper v1.7.0
	github.com/stretchr/testify v1.6.1
	github.com/tektoncd/cli v0.16.0
	github.com/tektoncd/pipeline v0.20.1
	go.uber.org/zap v1.16.0
	goa.design/goa/v3 v3.2.2
	goa.design/plugins/v3 v3.1.3
	golang.org/x/crypto v0.0.0-20201221181555-eec23a3978ad
	golang.org/x/mod v0.4.1 // indirect
	golang.org/x/oauth2 v0.0.0-20210126194326-f9ce19ea3013
	golang.org/x/term v0.0.0-20201210144234-2321bbc49cbf // indirect
	golang.org/x/time v0.0.0-20201208040808-7e3f01d25324 // indirect
	golang.org/x/tools v0.1.0 // indirect
	gopkg.in/h2non/gock.v1 v1.0.15
	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b // indirect
	gorm.io/driver/postgres v1.0.2
	gorm.io/gorm v1.20.7
	gotest.tools/v3 v3.0.2
	k8s.io/api v0.19.7
	k8s.io/apimachinery v0.19.7
	k8s.io/client-go v11.0.1-0.20190805182717-6502b5e7b1b5+incompatible
	k8s.io/utils v0.0.0-20210111153108-fddb29f9d009 // indirect
	knative.dev/pkg v0.0.0-20210203171706-6045ed499615
	maze.io/x/duration v0.0.0-20160924141736-faac084b6075
)

replace k8s.io/client-go => k8s.io/client-go v0.19.7
