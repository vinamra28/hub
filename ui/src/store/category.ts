import { types, getEnv, flow, Instance } from 'mobx-state-tree';
import { Api } from '../api';

export const Tag = types.model('Tags', {
  id: types.integer,
  name: types.string
});

export const Category = types
  .model('Category', {
    id: types.number,
    name: types.string,
    tags: types.array(Tag),
    selected: false
  })
  .actions((self) => ({
    toggle() {
      self.selected = !self.selected;
    }
  }));

export type ICategory = Instance<typeof Category>;
export type ITag = Instance<typeof Tag>;

export const CategoryStore = types
  .model('CategoryStore', {
    items: types.array(Category),
    isLoading: true,
    err: ''
  })

  .views((self) => ({
    get api(): Api {
      return getEnv(self).api;
    },

    get count() {
      return self.items.length;
    },

    get tags() {
      return self.items
        .filter((c) => c.selected)
        .reduce((acc: string[], c: ICategory) => [...acc, ...c.tags.map((t) => t.name)], []);
    }
  }))

  .actions((self) => ({
    add(item: ICategory) {
      self.items.push(item);
    },

    setLoading(l: boolean) {
      self.isLoading = l;
    },

    clear() {
      self.items.map((c: ICategory) => (c.selected = false));
    }
  }))

  .actions((self) => ({
    load: flow(function* () {
      try {
        self.setLoading(true);
        const { api } = self;
        const json = yield api.categories();
        json.data.forEach((item: ICategory) => self.add(item));
      } catch (err) {
        self.err = err.toString();
      }
      self.setLoading(false);
    })
  }))

  .actions((self) => ({
    afterCreate() {
      self.load();
    }
  }));

export type ICategoryStore = Instance<typeof CategoryStore>;
