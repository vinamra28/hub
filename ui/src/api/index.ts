import axios from 'axios';
import { API_URL } from '../config/constants';
import { ICategory } from '../store/category';
import { IResource } from '../store/resource';

export interface Api {
  categories(): Promise<ICategory>;
  resources(): Promise<IResource>;
}

export class Hub implements Api {
  async categories() {
    try {
      return axios.get(`${API_URL}/categories`).then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async resources() {
    try {
      return axios.get(`${API_URL}/resources`).then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async resourceVersion(resourceId: string) {
    try {
      return axios
        .get(`${API_URL}/resource/${resourceId}/versions`)
        .then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }
}
