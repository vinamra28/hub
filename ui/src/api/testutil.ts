import { Api } from './';
import * as fs from 'fs';
import { ICategory } from '../store/category';
import { IResource } from '../store/resources';

export class FakeHub implements Api {
  dataDir: string;

  constructor(dataDir: string) {
    this.dataDir = dataDir;
  }

  async resources() {
    const data = `${this.dataDir}/resources.json`;

    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<IResource>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }

  async categories() {
    const data = `${this.dataDir}/categories.json`;

    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<ICategory>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }
}
