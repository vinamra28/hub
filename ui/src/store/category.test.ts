import { when } from 'mobx';
import { getSnapshot } from 'mobx-state-tree';
import { FakeHub } from '../api/testutil';
import { CategoryStore, Category } from './category';

const TESTDATA_DIR = `${__dirname}/testdata`;
const api = new FakeHub(TESTDATA_DIR);

describe('Category Object', () => {
  it('can create a category object', () => {
    const category = Category.create({
      id: 1,
      name: 'test',
      tags: [
        {
          id: 1,
          name: 'test-category'
        }
      ]
    });

    expect(category.name).toBe('test');
    expect(category.id).toBe(1);
    expect(category.tags[0].name).toBe('test-category');
  });
});

describe('Store functions', () => {
  it('can create a store', (done) => {
    const store = CategoryStore.create({}, { api });
    expect(store.count).toBe(0);
    expect(store.isLoading).toBe(true);
    when(
      () => !store.isLoading,
      () => {
        expect(store.count).toBe(5);
        expect(store.isLoading).toBe(false);

        expect(store.list[0].id).toBe(1);
        expect(store.list[0].name).toBe('Build Tools');
        expect(getSnapshot(store)).toMatchSnapshot();

        done();
      }
    );
  });

  it('can toggle the selected category', (done) => {
    const store = CategoryStore.create({}, { api });
    expect(store.count).toBe(0);
    expect(store.isLoading).toBe(true);

    when(
      () => !store.isLoading,
      () => {
        expect(store.count).toBe(5);
        expect(store.isLoading).toBe(false);

        store.list[0].toggle();

        expect(store.list[0].selected).toBe(true);

        done();
      }
    );
  });

  it('can return the tags for the categories which are selected', (done) => {
    const store = CategoryStore.create({}, { api });
    expect(store.count).toBe(0);
    expect(store.isLoading).toBe(true);

    when(
      () => !store.isLoading,
      () => {
        expect(store.count).toBe(5);
        expect(store.isLoading).toBe(false);

        store.list[0].toggle();
        store.list[1].toggle();

        const tags = store.tags;
        expect(tags[0]).toBe('build-tool');

        done();
      }
    );
  });

  it('clears all the selected categories', (done) => {
    const store = CategoryStore.create({}, { api });
    expect(store.count).toBe(0);
    expect(store.isLoading).toBe(true);

    when(
      () => !store.isLoading,
      () => {
        expect(store.count).toBe(5);
        expect(store.isLoading).toBe(false);

        store.list[0].toggle();
        store.list[2].toggle();
        store.clear();

        expect(store.list).toEqual(
          expect.arrayContaining([
            expect.objectContaining({
              selected: false
            })
          ])
        );

        done();
      }
    );
  });
});
