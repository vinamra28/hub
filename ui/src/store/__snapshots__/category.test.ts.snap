// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store functions can create a store 1`] = `
Object {
  "err": "",
  "isLoading": false,
  "list": Array [
    Object {
      "id": 1,
      "name": "Build Tools",
      "selected": false,
      "tags": Array [
        Object {
          "id": 1,
          "name": "build-tool",
        },
      ],
    },
    Object {
      "id": 2,
      "name": "CLI",
      "selected": false,
      "tags": Array [
        Object {
          "id": 2,
          "name": "cli",
        },
      ],
    },
    Object {
      "id": 3,
      "name": "Cloud",
      "selected": false,
      "tags": Array [
        Object {
          "id": 4,
          "name": "aws",
        },
        Object {
          "id": 5,
          "name": "azure",
        },
        Object {
          "id": 6,
          "name": "cloud",
        },
        Object {
          "id": 3,
          "name": "gcp",
        },
      ],
    },
    Object {
      "id": 4,
      "name": "Deploy",
      "selected": false,
      "tags": Array [
        Object {
          "id": 7,
          "name": "deploy",
        },
      ],
    },
    Object {
      "id": 5,
      "name": "Image Build",
      "selected": false,
      "tags": Array [
        Object {
          "id": 8,
          "name": "image-build",
        },
      ],
    },
  ],
}
`;
